### ISSUE TITLE
cache selector results

### ISSUE THREAD

**Issue by xudifsd** [kubernetes/kubernetes#5548]
This is a proposal to implement caching selector result. https://github.com/GoogleCloudPlatform/kubernetes/issues/4817#issuecomment-76495394

AFAIK, currently only [two components](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/labels.md#api) use selector to group their members: `service` and `replicationController`, and they get list of their members by matching seletor against all pods' label every time.

Here I propose a way to cache seletor result in etcd.

Define a new registry in etcd: selector.

Whenever a `service` or `replicationController` is created, generate [uid](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/identifiers.md#uids) for its selector, and register selector with that uid in etcd. While creating `service` or `replicationController`, it watches that selector in etcd to wait updates on result of selector. We should add pre-existed matched pods to its result of course.

And whenever a `pod` is created, matches its label with all selector under its namespace (assuming we forbidden selector select object in another ns, which I not quite sure), and add that pod to selector result if match. Update or delete a `pod` should also update all selectors' result under its ns.

When a `selector` and `pod` is created at same time, we should use some kind of mutex to avoid race condition.

Any comment?

**Comment by jlowdermilk** [comment]
cc @thockin, @bgrant0607

**Comment by bgrant0607** [comment]
cc @bprashanth

**Comment by derekwaynecarr** [comment]
I think this complicates the delete life-cycle for any object that is or will be selector based.

Is there a reason that the cache cannot be held in memory rather than materialized in etcd directly?

**Comment by xudifsd** [comment]
@derekwaynecarr I just want to use etcd watch to do cache update, or are there any way for master to notify every minion that an object is updated?

**Comment by smarterclayton** [comment]
Minions are already watching the master for changes to pods.  What would you be adding a watch for here?

**Comment by xudifsd** [comment]
@smarterclayton I'm adding a selector result, not pod, in this way, `service` or `replicationController` can just watch selector result for changes, and do not need to compute selector result on the fly anymore.

**Comment by bgrant0607** [comment]
I was thinking we'd just implement the index in memory.

In particular, the scheduler, kube-proxy, replication controller manager, and endpoints controller could build the index on the client side.

The basic idea of matching against selectors in the same namespace upon pod creation/deletion is the right idea.

**Comment by bgrant0607** [comment]
It's also possible for clients to list or watch filtered by label selectors. For that, we would ideally have some sort of indexing in the server. For list, it's reasonably straightforward. For watch at a particular resourceVersion, it's admittedly harder.

**Comment by smarterclayton** [comment]
We do at least support watches filtered by labels, but it doesn't tell the user what the change was.  I've been toying with the idea of including the previous version in the watch for clients as well.

**Comment by xudifsd** [comment]
IMO, index is just used to accelerate the selector select, and it can be deployed on both server & client side. It is very useful for computing result on the fly. But to gain this benefit, we have to maintain the index upon object update, if we do index on the client side, all clients should do exact same update to its index (they need to maintain index for all objects under all namespaces)

@smarterclayton  isn't etcd watch already supported this feature?

```bash
curl "http://127.0.0.1:2379/v2/keys/somekey?wait=true"
{"action":"set","node":{"key":"/somekey","value":"someval1","modifiedIndex":10,"createdIndex":10},"prevNode":{"key":"/somekey","value":"someval","modifiedIndex":9,"createdIndex":9}}
```

After some thought on concerns by @derekwaynecarr , maybe we could use a go routine to do selector result update.

So, let me make it more clear:
This proposal create a new registry in etcd: `selector`, it has uid, requirement and list of matched objs attached. Selector is created once object like `service` & `replicationController` that has selector is created. We also need a go routine(let's call it `selectorManager`) to make things more modular, this routine is started by master at startup, it watches all the `pod`s & all the `selector`s, and update `selector` when matched pod is added or deleted, and traverse all `pod`s if a new `selector` is added, since we do this in one thread, we don't need mutex I mentioned before.

What changed to current system:
* whenever a selector based obj is deleted, we just need to delete the selector that belongs to it. This doesn't complicate delete life-cycle for selector based object much, right? @derekwaynecarr, this also applied to create life-cycle, since `selectorManager` is responsible for fill the data.
* selector based obj like `service` could just watches its `selector` for update.

@bgrant0607 why we need watch at a particular resourceVersion?

I want to do this as GSoC project. Because I'm not quite familiar with code base, so there must be some flaws in my proposal, but I want to make sure this is on the right direction, and we can update details later. Also since GSoC doesn't start until 25th May, and this project is fast-moving, I think this idea is appropriate as GSoC project: it's not core functionality, just optimization, and can be done in parallel with index.

What do you think?

**Comment by gmarek** [comment]
I'm slightly worried that this will roughly double the number of etcd reads that is required to get some interesting data. IIRC we were worried about etcd performance, but maybe something's changed.

New registry is probably better than current 'read everything' behavior, but I like Brians suggestion to do it in-memory client side way more. I'm much less worried of doing single in-memory update of the index in every client for every watched change than duplicating number of etcd reads.

As for deadline, I'm afraid that some kind of index or other way to improve selection performance need to be done way before May 25th...

**Comment by bprashanth** [comment]
To sum up the concerns in this thread, the idea of having a selector cache is useful, but one does not need to push something as an individual resource into etcd to be able to watch it. Say you had an rc for each of your 2 dbs, they could in theory do:

```
curl http://<apiserver>/api/<version>/watch/pods?labels=name=elasticsearch
curl http://<apiserver>/api/<version>/watch/pods?labels=name=mysql
```

Even though labels are not a resource in etcd, because the apiserver supports http long polling. Is there additional functionality you're proposing above this? If not, we should tackle this as a performance and scale optimization, and using a db as a cache has never helped with either :)

For performance you want to quickly match up the count of active pods that match a selector (perferably without having to serialize->wire transport->deserialize just to take a len() on the other side) and for scale you want it so we don't have 1 connection per rc. I feel like we can achieve both with the right interface for selectors, backed by the api we already have and an in memory index.

**Comment by derekwaynecarr** [comment]
I added client side caching that supports indexing a while ago:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/client/cache/store.go#L261

I think the scale-out problem is really limited with our current goals for 1.0, and the only areas I could see where we have a potential problem is with pods.  The pod queries are only problematic with respect to those that cross namespace boundaries or where we have a huge number of pods in a single namespace.

For replication controllers, we should look to modify replication_controller.go syncReplicationController to just do a LIST/WATCH on pods and index by namespace and by selector.  That would eliminate the per namespace query in each sync method.  We would still need to filter out failed/succeeded pods as we do now.  I guess I don't see a major benefit to introducing something in the API server versus just doing this in the replication controller.

I also have yet to see where replication controllers shows a performance problem to know if we are not taking on a lot of code to do premature optimization in this area.

**Comment by bprashanth** [comment]
We can't watch pods till we start syncing their status (or it might work because of some implementation detail of the pod cache but it won't be pretty), IIUC the change to sync pod status should be in soon. Once we have that we can essentially watch updates to pods/status. 

Currently, the rc has a poll loop with a totally artificial delay of 10s. This means if i kill pods of an rc, it will take that long before the rc notices. If I *resize* the rc it will get notified of changes immediately, because it watches. This also affects the performance of `rc stop`, because we update the Status.Replicas only every 10s. Ideally we would update it as pods went offline and kubectl would just watch the Status.Replicas field, but it currently polls the rc introducing another artificial delay of 3s (https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/stop.go#L78).

I also can't make a *hard* case for maintaing a more complex index on the server, however I'm wary of multiple selector caches popping up in various clients (i.e we need to be careful about not ending up with n pod caches). A cache on the server would also be useful for other reasons (eg: https://github.com/GoogleCloudPlatform/kubernetes/issues/2210), but I agree it might be premature for v1.

**Comment by smarterclayton** [comment]
There should only ever (probably) be one type of cache per resource.  The indexes on that cache should be defined.  Eventually, it should be possible to watch on the cache or the associated reflector (as well as do catch up) so that we don't create one etcd watch per API client.  To do this, we need to get better at making Registry interface match Client interface, and have consistent mapping of errors and behavior from Registry -> REST and Client -> REST, as well as and Registry -> Cache and Client -> Cache.  A lot of the pieces are in place though.

----- Original Message -----

**Comment by bprashanth** [comment]
I think there are 2 issues being discussed here and we should pare down to just 1 (@xudifsd 's gsoc proposal). I think we agree that pushing selectors into etcd sounds like a bad idea, and implementing a lookup/reverse lookup of labels -> resources in the apiserver is out of scope for 1.0. 

However, we don't have to commit right now to a caching technique across the board. In general I feel there are multiple candidates for a good cross-resource index (labels, hostip, status). It would be good to answer questions higher level applications might ask in constant time (eg all rcs, services, pods matching a label for a ui; all nodes running a canary build). So exploring an index/reverse-index that we maintain on the apiserver makes sense, but the spec needs to be fleshed out. That's part of the work involved.

**Comment by derekwaynecarr** [comment]
A mechanism to lookup by UID seems valuable long term and a current gap.

**Comment by gmarek** [comment]
I'm not certain that it's out of scope for 1.0. We're currently not fast enough for LIST operations, and some kind indexing is one of ideas how to make it better.

**Comment by xudifsd** [comment]
ok, it seems putting selector into etcd is a bad idea.

I was planning to use etcd because I don't know any other method for server to notify changes to selector result.

Since apiserver supports http long-polling like @bprashanth said, maybe we could do something like

```
curl http://<apiserver>/api/<version>/watch/selector_uid?version=10
```

and version is unique, monotonically-incrementing integer created for each change to selector. The long-polling returns once version of watched selector exceed version provided by client, and the result returned isn't incremental update on previous version, since it's too hard to keep every version, finally get what @bgrant0607 means by watch at a particular resourceVersion.

Would it be a problem since every client have a long-polling request to apiserver?

BTW, when do you plan to release v1.0?

**Comment by bprashanth** [comment]
@gmarek we can solve the perf problems with simpler client side lookups. LIST is fairly fast if you do it once, just not once per rc. Daniel's controller framework should help too.

**Comment by gmarek** [comment]
@bprashanth: Thanks, good to know. I didn't know what the current status is, and I remember Brendan saying that there's a problem with it.

**Comment by bprashanth** [comment]
@xudifsd:
* @lavalamp 's opinions on this would be indispensible, since he cryptically said https://github.com/GoogleCloudPlatform/kubernetes/issues/4817#issuecomment-76482950 and I know he has thoughts on a caching layer (i believe he's in next week)
* Don't think we want to start of by giving the selectors themselves a resource version

lets start with fleshing out the requirements before jumping into the implementation. You have the gsoc timeframe for that. 

We use labels as internal glue in our system and currently we need to repeatedly list and filter to figure out matching objects (linear search). One way to avoid this would be to have an index on the apiserver that watched etcd for changes to said objects (the same way it currently does), but creates an index, and uses it to answer questions through the same api it currently exposes. So your request would still look like:
```
curl http://<apiserver>/api/<version>/watch/<resource,eg: pods>?labels=name=foo&resourceVersion=<version of resource>
```

Only that every such query would *not* list pods from etcd and filter them. They would hit the index to lookup pods that match the label. The resources would obviously still have resourceVersions.

As described the index might as well be per resource and live on the client, but you also want to facilitate quick cross-resource lookup (https://github.com/GoogleCloudPlatform/kubernetes/issues/3676). It is unclear if such lookup will actually need some resourceVersion based grouping at all, we might get away with serving stale data directly from cache.

**Comment by derekwaynecarr** [comment]
I will add that I expect if I do

(1) GET /api/version/namespaces/foo/pods
(2) GET /api/version/namespaces/foo/pods?labels=...&fields=...

Both (1) and (2) should be served by the same cache and have the same staleness characteristic.

Generally, this means that the api server just holds everything in memory, and serves results from the same backing store for queries filtered by label or not.

Make sense?



----- Original Message -----

**Comment by gmarek** [comment]
I'm slightly worried by mixing terms 'index' and 'cache'. For me former means keeping indexed labels in memory and reading full data from etcd (I believe labels will be changing far less often than other 'content'), while the latter means keeping in-memory copy of whole etcd and serving results from it.

**Comment by derekwaynecarr** [comment]
I am for the latter, and we have a solution today to index the data in the latter form using cache.Store and cache.Indexer 

There is no call in etcd that lets you get N objects in a single request by passing N keys.

**Comment by xudifsd** [comment]
@bprashanth yeah, I'm also a little bit confused by index and cache, it makes more sense to me if you mean cache in your reply.

I guess I have to read more docs and code to have better understanding about requirements. Do you have any recommendation? It's a little scary to face 500k LOC directly

**Comment by bprashanth** [comment]
Cache.Indexer is a fine recommendation (https://github.com/GoogleCloudPlatform/kubernetes/blob/89f18f6fe1b541523e7f944f6f2ac45cfc0704fd/pkg/client/cache/store.go). You can also start basic, and understand how a watch fires on the apiserver (i.e why do we have the problem of repeated listing that's being discussed in this thread).

**Comment by xudifsd** [comment]
I guess it'll spend too much memory to cache all the pods in server side. Maybe it's reasonable just keep selector index in server's memory, and use it to serve selector select. Also maybe we could implement some watch mechanism to serve list & watch method may requested by client

**Comment by smarterclayton** [comment]
That depends - at the scales we are targeting in 1.0-1.2 it should be no problem.  Kube is targeted (right now) with having a cluster design <10k nodes.  Even at 50 pods per node 50k*1K objects is still only 50M.

We have strategies for sharding clusters of necessary, so we are still within the memory bounds for our primary focus.

**Comment by xudifsd** [comment]
@smarterclayton maybe you made a mistake on calculation: 50 * 10k * 1K = 500M memory, maybe this still isn't a problem?
But it seems we don't need to cache all the pods in memory at all, we can maintain index of it, and the index serve selector request, then we use index to get pods from etcd?

**Comment by smarterclayton** [comment]
> On Mar 22, 2015, at 10:06 PM, Di Xu <notifications@github.com> wrote:
> 
> @smarterclayton maybe you made a mistake on calculation: 50 * 10k * 1K = 500M memory, maybe this still isn't a problem?
> 
Lost the zero, but yes.  We've budgeted for that in previous designs, and although it's a lot of memory the goal is to try and find the design with the lowest complexity and the highest performance while still allowing us to scale clusters appropriately.
> But it seems we don't need to cache all the pods in memory at all, we can maintain index of it, and the index serve selector request, then we use index to get pods from etcd?
> 
Etcd also holds everything in memory today (planned to change in the future), but the plan so far has been to work on the assumption that our high end bound for a single Kube cluster can run from memory on the apiservers.

**Comment by derekwaynecarr** [comment]
I am not sure how much you know about etcd api, but there is no fetch nodes by keys operation.  You can get a single node by key, or all child nodes by key.  It's hard to follow your comment on what is meant by use index to get pods from etcd without more detail.

I see no issue holding cluster state in memory and I believe this is how Google runs internally today.  The current cache.Indexer in client package still seems fine to me at providing desired use case.

**Comment by xudifsd** [comment]
@derekwaynecarr yeah, I know that, so, I mean if we just keep index, then the index value maybe is a bunch of keys of pod, and if we need to fetch them, we have to fetch them one by one. But if we keep pods in memory, we can be much faster.

I'm trying to understand the constraints we currently have and finalize the proposal.

**Comment by xudifsd** [comment]
Hi, I just wrote up the draft of my GSoC [proposal](https://docs.google.com/document/d/1aF_Tf0xwV3NDeYtToyZxTp-cZZMsgbWorGl8QOr2INk/edit?usp=sharing)
some details need to be figured out, I'm wondering who will be my mentor if I get in GSoC, it's very hard for me to work this out on my own.

**Comment by bprashanth** [comment]
@xudifsd looks good! I've never seen a gsoc proposal before, someone who has (@bgrant0607 ?) might be able to gauge. You can always flesh it out when you start.

@derekwaynecarr the only danger with doing it all on the clients is fragmentation and creeping normality. Each cache is bound of have different lag/consistency guarantees which might affect the manageability of the system. Not saying building a consistent, race-free index atop etcd is going to be bug free, but I get the feeling (based solely off intuition) that it'll be easier to debug.

**Comment by bgrant0607** [comment]
@xudifsd The goal is for 1.0 to land sometime in June.

There are multiple problems discussed on this issue.

One problem is that currently controllers and the scheduler are doing a lot of redundant work. 

For instance, the replication controller manager lists pods for every RC:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/controller/replication_controller.go#L195

For each pod, the scheduler lists all services:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/scheduler/predicates.go#L275
then for each service targeting the pod, it lists all pods:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/scheduler/predicates.go#L280
and it walks all pods for each node:
https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/scheduler/predicates.go#L357

We'll likely need to reduce the number of queries to the apiserver in order to meet our performance goals. One simple approach discussed for the replication controller manager would be to get all pods once, then update the counts of all replication controllers. Narrowing the search by namespace may be sufficient to identify the matching RCs for now.

Eventually, such clients will want to continuously watch all objects of interest (RCs and pods for replication controller manager, and at least nodes, pods, and services for the scheduler) and then incrementally update their state as objects are created, updated, and deleted, in order to avoid repeated listing entirely. This will require finding related objects in constant time, which implies linking them on the client side. Also note that incremental updates requires implementation of efficient reverse lookups, to identify which selectors' results need to be updated when an object is mutated.

I also want to make it feasible to push lookups for more detailed information to clients (e.g., which services target a pod), as discussed in #3676, which is only feasible if the call pattern doesn't incur quadratic complexity (or worse). This would require indexing/linking on the server, which would be used to serve at least list queries. 

Therefore, I think we'll need to pursue both client-side and server-side approaches.

It's likely this problem won't be completely solved by May. Whatever remains (server-side indexing, efficient watch filtering, reverse lookups, index by uid or ip address, etc.) should be enough to comprise a project.

I recommend thinking about how to more efficiently solve the reverse lookup problem -- how to quickly find a matching selector, especially when using the generalized selectors (#341).

**Comment by bprashanth** [comment]
Fyi, more of an implementation detail but the lists in the scheduler should be hitting local stores populated by a reflector: https://github.com/GoogleCloudPlatform/kubernetes/blob/master/plugin/pkg/scheduler/factory/factory.go#L145

**Comment by xudifsd** [comment]
@bgrant0607 @bprashanth Thanks for the pointer, I'll try to follow the progress of kubernetes, and familiar with code base, submit some PRs if possible.

Current issue in the proposal is I can't give any detail of solution or anything like that, because the situation might be different when GSoC start.

P.S. what's the best way to ask question about the code? [mail list](https://groups.google.com/forum/#!forum/kubernetes-dev) seems not very active.

**Comment by bgrant0607** [comment]
IRC: #google-containers on freenode is the best way to get help.

**Comment by bgrant0607** [comment]
cc @lavalamp @quinton-hoole

**Comment by bgrant0607** [comment]
Good use case for caching/indexing in the apiserver: Kubelets watching the pods assigned to them. Currently, every Kubelet effectively lists/watches all pods.

**Comment by lavalamp** [comment]
OK, it looks like you guys figured out without me that storing an index in etcd is not tenable. :)

I will add a design to #4817 tomorrow.

**Comment by smarterclayton** [comment]
I don't remember linking this here, but Openshift uses cache.Indexer to reverse index services by portal IP for DNS.  

https://github.com/openshift/origin/blob/master/pkg/dns/serviceaccessor.go

**Comment by lavalamp** [comment]
See: https://github.com/GoogleCloudPlatform/kubernetes/issues/4817#issuecomment-89437368

**Comment by gmarek** [comment]
I think we can this issue. We have API-server side cache now.
